src/context/ModalProvider/ModalProvider.tsx

```diff
@@ -1,116 +1,130 @@
+type ModalProps = {
+  [k in Modals]: React.ComponentProps<typeof MODALS[k]['component']>
 }
 
+type ModalActions = OpenModalType<Modals> | CloseModalType
+
 // reducer
 const reducer = (state: ModalState, action: ModalActions) => {
   switch (action.type) {
     case OPEN_MODAL:
-      validateName(action.name)
-      return { ...state, [action.name]: true, props: action.props }
+      return {
+        ...state,
+        [action.name]: { ...state[action.name], isOpen: true, props: action.props }
+      }
     case CLOSE_MODAL:
-      validateName(action.name)
-      return { ...state, [action.name]: false, props: {} }
+      return { ...state, [action.name]: { ...state[action.name], isOpen: false } }
     default:
       return state
   }
 }
 
 // state
-const initialState = Object.freeze({
-  [MODALS.SEND]: false,
-  [MODALS.RECEIVE]: false,
-  props: {}
-})
+const initialState = Object.freeze(MODALS)
 
 // context
-export const ModalContext = React.createContext<ModalContextProps>({
-  ...initialState,
-  close: noop,
-  open: noop
-})
+export const ModalContext = React.createContext(initialState)
 
 // provider
 export const ModalProvider = ({ children }: ModalProviderProps) => {
   const [state, dispatch] = useReducer(reducer, initialState)
 
-  const open = useCallback(
-    (name: Modal, props?: AllowedProps) =>
-      dispatch({
-        type: OPEN_MODAL,
-        name,
-        props: props ?? ({} as AllowedProps)
-      }),
-    [dispatch]
-  )
-
-  const close = useCallback(
-    (name: Modal) =>
-      dispatch({
-        type: CLOSE_MODAL,
-        name
-      }),
-    [dispatch]
-  )
-
-  const value = useMemo(() => ({ ...state, close, open }), [close, open, state])
+  const openFactory = useMemo(<K extends Modals>() => {
+    return (name: Modals) => (props: ModalProps[K]) => dispatch({ type: OPEN_MODAL, name, props })
+  }, [])
+
+  const closeFactory = useMemo(() => {
+    return (name: Modals) => () => dispatch({ type: CLOSE_MODAL, name })
+  }, [])
+
+  const value = useMemo(() => {
+    const modals = Object.values(MODALS).map(modal => modal.name)
+    const fns = modals.reduce((acc, cur) => {
+      const open = openFactory(cur)
+      const close = closeFactory(cur)
+      return { ...acc, [cur]: { open, close } }
+    }, state)
+    const result = merge(state, fns)
+    return result
+  }, [state, closeFactory, openFactory])
 
   return (
     <ModalContext.Provider value={value}>
       {children}
-      <SendModal />
-      <ReceiveModal />
+      {/*
+          if someone can come up with a way to dynamically
+          narrow the type of Modals here without hardcoding
+          each modal or explicitly specifying types i'd love to see it
+      */}
+      {/* we can safely non null assert here as the props are saved to value (state) on the open action */}
+      {value.send.isOpen && <value.send.component {...value.send.props!} />}
+      {value.receive.isOpen && <value.receive.component {...value.receive.props!} />}
```

Comment:
Could you do something along the lines of this?
```
      {Object.values(MODALS).map(modal => {
        return modal.isOpen && <modal.component key={modal.name} {...modal.props} />
      })}
```

Reply:
yup, therein lies the problem - at runtime you can't dynamically narrow the type of `modal` when mapping from `'send' | 'receive'` to a concrete type, so components and props become a union and it expects all props to be assignable to all components which is wrong.

believe me i tried 😖

Reply:
Ah I see yea that might be a though one.